src/cargo/ops/cargo_new.rs

```diff
@@ -85,6 +97,87 @@ struct CargoNewConfig {
     version_control: Option<VersionControl>,
 }
 
+enum TemplateDirectory{
+    Temp(TempDir),
+    Normal(PathBuf)
+}
+
+impl TemplateDirectory {
+    fn path(&self) -> &Path {
+        match *self {
+            TemplateDirectory::Temp(ref tempdir) => tempdir.path(),
+            TemplateDirectory::Normal(ref path) => path.as_path()
+        }
+    }
+}
+
+// The type of template we will use.
+#[derive(Debug, Eq, PartialEq)]
+enum TemplateType<'a>  {
+    GitRepo(&'a str),
+    LocalDir(&'a str),
+    Builtin
+}
+
+fn get_template_type<'a>(repo: Option<&'a str>, 
+                         name: Option<&'a str>
+                         ) -> CargoResult<TemplateType<'a>> {
+    match (repo, name) {
+        (Some(repo_url), _) if repo_url.starts_with("http://")  ||
+            repo_url.starts_with("https://") ||
+            repo_url.starts_with("file://") ||
+            repo_url.starts_with("git@") => Ok(TemplateType::GitRepo(repo_url)),
+        (Some(directory), _) => Ok(TemplateType::LocalDir(directory)),
+        (None, Some(_)) => Err(human("A template was given, but no template repository")),
+        (None, None) => Ok(TemplateType::Builtin)
+    }
+}
+
+fn get_input_template(config: &Config, opts: &MkOptions) -> CargoResult<TemplateDirectory> {
+    let path = opts.path;
+    let name = opts.name;
+    let templates_dir = config.template_path();
+    if fs::metadata(&templates_dir).is_err() {
+        try!(fs::create_dir_all(&templates_dir));
+    }
+
+    let template_type = try!(get_template_type(opts.template_repo, opts.template));
+    let template_dir = match template_type {
+        // given template is a remote git repository & needs to be cloned
+        TemplateType::GitRepo(repo_url) => {
+            let template_dir = try!(TempDir::new(name));
+            println!("Checking out repo to dir: {}", template_dir.path().display());
+
+            try!(config.shell().status("Cloning", repo_url));
+            match Repository::clone(repo_url, &*template_dir.path()) {
```

Comment:
Could this use the support in the `sources::git` module? That way it'd handle cloning concerns like proxies and such.